### PR TITLE
tbbbind support CC

### PR DESCRIPTION
**Pull request by riverlijunjie**
### Details:
 - tbbbind support conditional compilation

### Tickets:
 - 70958

### CODE REVIEW

**Comment by dmitry-gorokhov** [comment]
@riverlijunjie Do you have numbers that demonstrate binary size reduction in CC scenario with this feature enabled?

**Comment by dmitry-gorokhov** [comment]
@ilya-lavrenov FYI

**Review by ilya-lavrenov — APPROVED** [review]

**Comment by ilya-lavrenov** [comment]
CC @ilyachur

**Comment by riverlijunjie** [comment]
> @riverlijunjie Do you have numbers that demonstrate binary size reduction in CC scenario with this feature enabled?

Not compare the size, but I don't expect there will be obvious size reduction due to tbbbind itself very small.

**Comment by dmitry-gorokhov** [comment]
@riverlijunjie Do you have any updates?

File: src/inference/src/threading/itt.hpp [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+ * @file itt.hpp
+ */
+
+#pragma once
+
+#include <openvino/cc/selective_build.h>
+
+#include <openvino/itt.hpp>
+
+namespace InferenceEngine {
```
Comment (ilyachur) [inline src/inference/src/threading/itt.hpp:16]: can we use ov namespace for new features?

**Comment by ilya-lavrenov** [comment]
@riverlijunjie Do you have any updates?

**Comment by riverlijunjie** [comment]
> 

There are 7 apis in tbbbinder, I verified and found 4 apis are called in test sample but all 7 apis still added into openvino lib. I didn't have time to investigate more. So let's keep this PR open until investigate is done.

**Comment by riverlijunjie** [comment]
It is very strange that after enable tbbbind CC, we found that only 3 tbbbind functions are called, but all 7 tbbbind functions will be built into libopenvino.so, it cause the libopenvino.so size has no change.

$ objdump -d libopenvino.so | grep __TBB_internal | grep callq
  644dd8:       e8 d3 cd 1f 00          callq  841bb0 <__TBB_internal_destroy_system_topology>
  644e36:       e8 95 ca 1f 00          callq  8418d0 <__TBB_internal_deallocate_binding_handler>
  644e96:       e8 35 ca 1f 00          callq  8418d0 <__TBB_internal_deallocate_binding_handler>
  644f65:       e8 46 cd 1f 00          callq  841cb0 <__TBB_internal_initialize_system_topology>

$ objdump -d libopenvino.so | grep __TBB_internal | grep ">:"
0000000000841620 <__TBB_internal_allocate_binding_handler>:
00000000008418d0 <__TBB_internal_deallocate_binding_handler>:
0000000000841950 <__TBB_internal_apply_affinity>:
00000000008419c0 <__TBB_internal_restore_affinity>:
0000000000841a00 <__TBB_internal_get_default_concurrency>:
0000000000841bb0 <__TBB_internal_destroy_system_topology>:
0000000000841cb0 <__TBB_internal_initialize_system_topology>:

Notice: tbbbind only is applied in TBB scenario,  while oneTBB will not use tbbbind so this PR will is disabled in oneTBB case.

**Comment by ilya-lavrenov** [comment]
> Notice: tbbbind only is applied in TBB scenario, while oneTBB will not use tbbbind so this PR will is disabled in oneTBB case.

OneTBB has TBBBind loaded in runtime. While for versions older than 2021 we have statically compiled TBBBind.

**Comment by riverlijunjie** [comment]
> > Notice: tbbbind only is applied in TBB scenario, while oneTBB will not use tbbbind so this PR will is disabled in oneTBB case.
> 
> OneTBB has TBBBind loaded in runtime. While for versions older than 2021 we have statically compiled TBBBind.

yes, but it seems that tbbbind CC only strip unused tbbbind functions from openvino functions,  but didn't remove the unused tbbbind functions from openvino binaries.
I have write a simple test sample for tbbbinder:

`#include <cstdlib>
#include <iostream>

class binding_handler;

extern "C" {
void __TBB_internal_initialize_system_topology(std::size_t groups_num,
                                               int& numa_nodes_count,
                                               int*& numa_indexes_list,
                                               int& core_types_count,
                                               int*& core_types_indexes_list);
void __TBB_internal_destroy_system_topology();
binding_handler* __TBB_internal_allocate_binding_handler(int number_of_slots,
                                                         int numa_id,
                                                         int core_type_id,
                                                         int max_threads_per_core);
void __TBB_internal_deallocate_binding_handler(binding_handler* handler_ptr);
void __TBB_internal_apply_affinity(binding_handler* handler_ptr, int slot_num);
void __TBB_internal_restore_affinity(binding_handler* handler_ptr, int slot_num);
int __TBB_internal_get_default_concurrency(int numa_id, int core_type_id, int max_threads_per_core);
}


void test(){
     std::size_t groups_num = 1;
     int numa_nodes_count = 1;
     int* numa_indexes_list = nullptr; 
     int core_types_count = 8;
     int* core_types_indexes_list = nullptr;

     __TBB_internal_initialize_system_topology(groups_num,
                                               numa_nodes_count,
                                               numa_indexes_list,
                                               core_types_count,
                                               core_types_indexes_list);

    binding_handler *hander = __TBB_internal_allocate_binding_handler(1,0,1,2);
    __TBB_internal_deallocate_binding_handler(hander);
    std::cout << "Test is executed" << std::endl;
    __TBB_internal_destroy_system_topology();
}

int main(int argc, char **argv) {
   test();
   return 1;
}`

Build it: `g++ -flto -fdata-sections -ffunction-sections main.cpp -o test -Wl,--gc-sections -L. -ltbbbind_2_5 -lhwloc`

We still see that 4 tbbbinder functions are called, but all 7 tbbbinder functions are linked into this binaries.

$ objdump -d test | grep __TBB_ | grep callq
    3834:       e8 b7 09 00 00          callq  41f0 <__TBB_internal_initialize_system_topology>
    384d:       e8 0e 03 00 00          callq  3b60 <__TBB_internal_allocate_binding_handler>
    385d:       e8 ae 05 00 00          callq  3e10 <__TBB_internal_deallocate_binding_handler>
    388d:       e8 5e 08 00 00          callq  40f0 <__TBB_internal_destroy_system_topology>

$ objdump -d test | grep __TBB_ | grep ">:"
0000000000003b60 <__TBB_internal_allocate_binding_handler>:
0000000000003e10 <__TBB_internal_deallocate_binding_handler>:
0000000000003e90 <__TBB_internal_apply_affinity>:
0000000000003f00 <__TBB_internal_restore_affinity>:
0000000000003f40 <__TBB_internal_get_default_concurrency>:
00000000000040f0 <__TBB_internal_destroy_system_topology>:
00000000000041f0 <__TBB_internal_initialize_system_topology>:

**Comment by riverlijunjie** [comment]
This PR need get:
    1. With NUMA some TBBBind API is optimized out, but some of the API is definitely here
    2. The same for HYBRID cores on Alder lake
    3. For no NUMA nor HYBRID, no TBBBind API is used

Currently, item 3 has been verified in my local machine, and WIP to get NUMA machine and ADL machine to verify item 1 and item 2.

**Review by ilyachur — APPROVED** [review]

**Review by wangleis — APPROVED** [review]

**Pull request merged**